## Librerias

In [1]:
import pandas as pd

import urllib.request
from PIL import Image

## Data

In [12]:
### Para que funcione necesitas bajarte los archivos de datos de Kaggle 
train = pd.read_csv("Dataset/tran.csv")


## Data exploration

In [3]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 912 entries, 0 to 911
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   laptop_ID         912 non-null    int64  
 1   Company           912 non-null    object 
 2   Product           912 non-null    object 
 3   TypeName          912 non-null    object 
 4   Inches            912 non-null    float64
 5   ScreenResolution  912 non-null    object 
 6   Cpu               912 non-null    object 
 7   Ram               912 non-null    object 
 8   Memory            912 non-null    object 
 9   Gpu               912 non-null    object 
 10  OpSys             912 non-null    object 
 11  Weight            912 non-null    object 
 12  Price_euros       912 non-null    float64
dtypes: float64(2), int64(1), object(10)
memory usage: 92.8+ KB


In [4]:
train.head()

laptop_ID Company             Product            TypeName  Inches  \
0       1223    Dell       Inspiron 5567            Notebook    15.6   
1         78  Lenovo  IdeaPad 320-15IKBN            Notebook    15.6   
2       1267    Dell              XPS 13  2 in 1 Convertible    13.3   
3        161    Dell       Inspiron 5579  2 in 1 Convertible    15.6   
4        922      LG         Gram 14Z970           Ultrabook    14.0   

                            ScreenResolution                         Cpu  Ram  \
0                          Full HD 1920x1080  Intel Core i5 7200U 2.5GHz  8GB   
1                          Full HD 1920x1080  Intel Core i5 7200U 2.5GHz  8GB   
2           Quad HD+ / Touchscreen 3200x1800   Intel Core i5 7Y54 1.2GHz  8GB   
3            Full HD / Touchscreen 1920x1080  Intel Core i7 8550U 1.8GHz  8GB   
4  IPS Panel Full HD / Touchscreen 1920x1080  Intel Core i7 7500U 2.7GHz  8GB   

      Memory                     Gpu       OpSys  Weight  Price_euros  
0  256GB SSD      AMD Radeon R7 M445  Windows 10  2.36kg        889.0  
1    2TB HDD   Intel HD Graphics 620       No OS   2.2kg        519.0  
2  256GB SSD   Intel HD Graphics 615  Windows 10  1.24kg       1813.0  
3  256GB SSD  Intel UHD Graphics 620  Windows 10  1.56kg       1049.0  
4  512GB SSD   Intel HD Graphics 620  Windows 10  0.98kg       1899.0

## Data processing

In [13]:
train = train[['laptop_ID','Inches','Price_euros']]

In [14]:
train

laptop_ID  Inches  Price_euros
0         1223    15.6       889.00
1           78    15.6       519.00
2         1267    13.3      1813.00
3          161    15.6      1049.00
4          922    14.0      1899.00
..         ...     ...          ...
907        709    15.6       399.00
908        947    13.3      1799.00
909        590    15.6      1537.39
910       1143    15.6       989.00
911        531    17.3       889.00

[912 rows x 3 columns]

In [ ]:
# df.to_csv("Dataset/limpio.csv")

-----------------------------------------------------------------------------------------------------------------

## Pasos que si o si debéis realizar para poder participar en la competición:

### 1. Definir X e y

In [15]:
X = train.drop(['laptop_ID','Price_euros'], axis=1)
y = train['Price_euros'].copy()
X.shape

(912, 1)

In [16]:
y.shape

(912,)

### 2. Dividir X_train, X_test, y_train, y_test

In [17]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 42)

In [18]:
X_train

Inches
25     14.0
84     15.6
10     15.6
342    15.6
890    13.3
..      ...
106    15.6
270    15.6
860    14.0
435    15.6
102    17.3

[729 rows x 1 columns]

### 3. Asignar el modelo (vacío) a una variable


In [19]:
from sklearn.linear_model import LinearRegression

regressor = LinearRegression()

regressor.fit(X_train, y_train)

LinearRegression()

### 4. Sacar métricas, valorar el modelo 

Recuerda que en la competición se va a evaluar con la métrica de RMSE.

In [20]:
from sklearn.metrics import root_mean_squared_error

predictions = regressor.predict(X_test)

rmse = root_mean_squared_error(y_test, predictions)

In [21]:
print(rmse)

756.144356190001


-----------------------------------------------------------------

## Una vez listo el modelo, toca predecir con el dataset de predicción 

Definición de **modelo que está listo**. 

Tras hacer suficientes pruebas, analizar los datos, hacer feature engineering, probar diferentes modelos con diferentes parámetros, es con este con el que observo mejores métricas y menos overfitting. ¡Cuidado con el overfitting aquí! Si vuestro modelo aprende muy bien de estos datos pero hay overfitting cuando le pasemos los datos desconocidos de `test.csv` nos arriesgamos a que digamos, no salga lo esperado.

### 1. Entrena dicho modelo con TODOS tus datos de train, esto es con `train.csv` al completo.


**CON LAS TRANSFORMACIONES QUE LE HAYAS REALIZADO A `X` INCLUÍDAS.**


Véase:
- Estandarización/Normalización
- Eliminación de Outliers
- Eliminación de columnas
- Creación de columnas nuevas
- Gestión de valores nulos
- Y un largo etcétera de técnicas que como Data Scientist hayas considerado las mejores para tu dataset.

### 2. Carga los datos de `test.csv` para predecir.

In [23]:
X_pred = pd.read_csv("Dataset/test.csv")
X_pred.head()

laptop_ID   Company                 Product   TypeName  Inches  \
0        539      Asus  Zenbook UX510UW-FI095T   Notebook    15.6   
1        327      Asus  ZenBook UX410UA-GV183T   Notebook    14.0   
2        563  Mediacom           SmartBook 130   Notebook    13.3   
3         13     Apple             MacBook Pro  Ultrabook    15.4   
4        935        HP           EliteBook 850  Ultrabook    15.6   

                     ScreenResolution                          Cpu   Ram  \
0     IPS Panel 4K Ultra HD 3840x2160   Intel Core i7 7500U 2.7GHz   8GB   
1                   Full HD 1920x1080   Intel Core i7 7500U 2.7GHz   8GB   
2         IPS Panel Full HD 1920x1080  Intel Atom x5-Z8350 1.44GHz   4GB   
3  IPS Panel Retina Display 2880x1800         Intel Core i7 2.8GHz  16GB   
4                   Full HD 1920x1080   Intel Core i7 6500U 2.5GHz   8GB   

                 Memory                      Gpu       OpSys  Weight  
0  256GB SSD +  1TB HDD  Nvidia GeForce GTX 960M  Windows 10     2kg  
1             256GB SSD    Intel HD Graphics 620  Windows 10     2kg  
2    32GB Flash Storage        Intel HD Graphics  Windows 10  1.35kg  
3             256GB SSD       AMD Radeon Pro 555       macOS  1.83kg  
4             256GB SSD      AMD Radeon R7 M365X  Windows 10  1.84kg

In [24]:
X_pred.shape

(391, 12)

**¿Por qué da error?** 

In [25]:
predictions_submit = regressor.predict(X_pred)
predictions_submit

ValueError: The feature names should match those that were passed during fit.
Feature names unseen at fit time:
- Company
- Cpu
- Gpu
- Memory
- OpSys
- ...


IMPORTANTE: APLICAR LO MISMO A ESTOS DATOS QUE HAYÁIS APLICADO A LOS DATOS DE ENTRENAMIENTO

- SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 4 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS
- SI AL ARRAY CON EL QUE HICISTEIS `.fit()` LO NORMALIZASTEIS, PARA `.predict()` DEBÉIS NORMALIZARLO
- TODO IGUAL SALVO BORRAR FILAS, EL NÚMERO DE ROWS SE DEBE MANTENER EN ESTE SET, PUES LA PREDICCIÓN DEBE TENER 391 FILAS, SI O SI

**Entonces, si al cargar los datos de train usé `index_col=0` para que utilizara la primera columna del conjunto de datos como índice, ¿tendré que hacerlo también para el conjunto `test.csv`?**

In [ ]:
# ¿Qué opináis?
# Sí

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

In [26]:
X_pred = pd.read_csv("Dataset/test.csv")
X_pred.head()

laptop_ID   Company                 Product   TypeName  Inches  \
0        539      Asus  Zenbook UX510UW-FI095T   Notebook    15.6   
1        327      Asus  ZenBook UX410UA-GV183T   Notebook    14.0   
2        563  Mediacom           SmartBook 130   Notebook    13.3   
3         13     Apple             MacBook Pro  Ultrabook    15.4   
4        935        HP           EliteBook 850  Ultrabook    15.6   

                     ScreenResolution                          Cpu   Ram  \
0     IPS Panel 4K Ultra HD 3840x2160   Intel Core i7 7500U 2.7GHz   8GB   
1                   Full HD 1920x1080   Intel Core i7 7500U 2.7GHz   8GB   
2         IPS Panel Full HD 1920x1080  Intel Atom x5-Z8350 1.44GHz   4GB   
3  IPS Panel Retina Display 2880x1800         Intel Core i7 2.8GHz  16GB   
4                   Full HD 1920x1080   Intel Core i7 6500U 2.5GHz   8GB   

                 Memory                      Gpu       OpSys  Weight  
0  256GB SSD +  1TB HDD  Nvidia GeForce GTX 960M  Windows 10     2kg  
1             256GB SSD    Intel HD Graphics 620  Windows 10     2kg  
2    32GB Flash Storage        Intel HD Graphics  Windows 10  1.35kg  
3             256GB SSD       AMD Radeon Pro 555       macOS  1.83kg  
4             256GB SSD      AMD Radeon R7 M365X  Windows 10  1.84kg

In [27]:
X_pred = X_pred[['Inches']]
X_pred

Inches
0      15.6
1      14.0
2      13.3
3      15.4
4      15.6
..      ...
386    13.3
387    13.3
388    15.6
389    14.0
390    15.6

[391 rows x 1 columns]

### 3. AHORA puedo hacer la predicción que será lo que subirás a Kaggle. 

**¿Qué es lo que subirás a Kaggle?**

In [28]:
predictions_submit = regressor.predict(X_pred)
predictions_submit

array([1121.40247869, 1098.40295269, 1088.34066007, 1118.52753794,
       1121.40247869, 1121.40247869, 1145.83947506, 1121.40247869,
       1121.40247869, 1098.40295269, 1121.40247869, 1098.40295269,
       1121.40247869, 1121.40247869, 1145.83947506, 1145.83947506,
       1063.9036637 , 1121.40247869, 1121.40247869, 1121.40247869,
       1145.83947506, 1088.34066007, 1145.83947506, 1121.40247869,
       1088.34066007, 1121.40247869, 1121.40247869, 1121.40247869,
       1098.40295269, 1145.83947506, 1096.96548232, 1088.34066007,
       1121.40247869, 1088.34066007, 1121.40247869, 1098.40295269,
       1145.83947506, 1145.83947506, 1121.40247869, 1098.40295269,
       1121.40247869, 1098.40295269, 1091.21560082, 1121.40247869,
       1121.40247869, 1121.40247869, 1121.40247869, 1145.83947506,
       1121.40247869, 1098.40295269, 1121.40247869, 1121.40247869,
       1121.40247869, 1121.40247869, 1121.40247869, 1098.40295269,
       1145.83947506, 1145.83947506, 1145.83947506, 1098.40295

**¡PERO! Para subir a Kaggle la predicción, ésta tendrá que tener una forma específica y no valdrá otra.**

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [29]:
sample = pd.read_csv("Dataset/sample_submission.csv")

In [30]:
sample.head()

laptop_ID  Price_euros
0        539          650
1        327          650
2        563          650
3         13          650
4        935          500

In [25]:
sample.shape

(391, 2)

### 4. Mete tus predicciones en un dataframe. 

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [33]:
submission = pd.DataFrame({"laptop_ID": sample['laptop_ID'], "Price_euros": predictions_submit})

In [34]:
submission.head()

laptop_ID  Price_euros
0        539  1121.402479
1        327  1098.402953
2        563  1088.340660
3         13  1118.527538
4        935  1121.402479

In [35]:
submission.shape

(391, 2)

### 5. Pásale el CHEQUEATOR para comprobar que efectivamente está listo para subir a Kaggle.

In [40]:
def chequeator(df_to_submit,version):
    """
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.
    """
    if df_to_submit.shape == sample.shape:
        if df_to_submit.columns.all() == sample.columns.all():
            if df_to_submit['laptop_ID'].all() == sample['laptop_ID'].all():
                print("You're ready to submit!")
                submission.to_csv(f"submission_{version}.csv", index = False) #muy importante el index = False
                urllib.request.urlretrieve("https://i.kym-cdn.com/photos/images/facebook/000/747/556/27a.jpg", "gfg.png")     
                img = Image.open("gfg.png")
                img.show()   
            else:
                print("Check the ids and try again")
        else:
            print("Check the names of the columns and try again")
    else:
        print("Check the number of rows and/or columns and try again")
        print("\nMensaje secreto de Clara: No me puedo creer que después de todo este notebook hayas hecho algún cambio en las filas de `diamonds_test.csv`. Lloro.")


In [41]:
chequeator(submission,1)

You're ready to submit!
